File: packages/python/policybook/metrics.py
"""Shared metric helpers.

Benchmark numbers are committed and compared across languages, so rounding has
to be specified rather than assumed.
"""

from __future__ import annotations

import math

__all__ = ["round6"]


def round6(value: float) -> float:
    """Round to six decimal places, half away from zero.

    Deliberately **not** Python's built-in ``round``, which uses banker's
    rounding: ``round(0.0000005, 6)`` gives 0.0 while JavaScript's
    ``Math.round(0.5) / 1e6`` gives 1e-06. The registry's reference
    implementation is JavaScript, so Python matches it with an explicit
    floor-of-plus-a-half.

    All registry metrics are non-negative, so there is no negative-tie case.
    """
    if not math.isfinite(value):
        return value
    return math.floor(value * 1e6 + 0.5) / 1e6

File: packages/core/src/metrics.ts
/**
 * Shared metric helpers.
 *
 * Benchmark numbers are committed and compared across languages, so rounding
 * has to be specified rather than assumed.
 */

/**
 * Round to six decimal places, half away from zero.
 *
 * Six places is enough to separate policies on a million-event trace without
 * recording noise. The rounding mode matters for ports: Python's built-in
 * `round` and C's `rint` are half-to-even and would disagree on an exact tie,
 * so all three languages spell it as `floor(value * 1e6 + 0.5) / 1e6`. Not
 * `Math.round`: for exactly one double (0.49999999999999994) the addition
 * rounds up to 1.0 where `Math.round` answers 0, so the two are *almost*
 * interchangeable — and "almost" is the wrong word to have inside a
 * bit-exactness contract. The reference implements the portable spelling.
 *
 * All registry metrics are non-negative, so there is no negative-tie case to
 * argue about.
 */
export function round6(value: number): number {
  if (!Number.isFinite(value)) return value;
  return Math.floor(value * 1e6 + 0.5) / 1e6;
}
